src/codegen/compile.ml

```diff
@@ -3865,11 +3880,89 @@ module StableMem = struct
           G.i (Binary (Wasm.Values.I64 I64Op.Add)) ^^
           compile_const_64 (Int64.of_int page_size_bits) ^^
           G.i (Binary (Wasm.Values.I64 I64Op.ShrU)) ^^
-          G.i (Convert (Wasm.Values.I32 I64Op.WrapI64)) ^^ (* TBR *)
+          G.i (Convert (Wasm.Values.I32 I64Op.WrapI64)) ^^
           compile_add_const 1l ^^
-          grow env)
+          ensure_pages env ^^
+          (* Check result *)
+          compile_unboxed_zero ^^
+          G.i (Compare (Wasm.Values.I32 I32Op.LtS)) ^^
+          E.then_trap_with env "Out of stable memory.")
+    | _ -> assert false
+
+  (* API *)
+
+  let logical_grow env =
+    match E.mode env with
+    | Flags.ICMode | Flags.RefMode ->
+      Func.share_code1 env "__stablemem_logical_grow"
+        ("pages", I32Type) [I32Type] (fun env get_pages ->
+          let (set_size, get_size) = new_local env "size" in
+          get_mem_size env ^^
+          set_size ^^
+
+          get_size ^^ G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32)) ^^
+          get_pages ^^ G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32)) ^^
+          G.i (Binary (Wasm.Values.I64 I32Op.Add)) ^^
+          compile_const_64 65536L ^^
+          G.i (Compare (Wasm.Values.I64 I64Op.LtU)) ^^
+          G.if_ []
+            (G.nop) (* fallthrough *)
```

Comment:
Nit: If you invert the condition (`GeU`) then you can use an else-less if and drop the nop.

Reply:
Wasn't aware we have that!

Reply:
I don’t think our `G.` module provides else-less ifs. But the wasm encoder might be smart here.

Reply:
Rewritten as full if, inverting test, and pushing continuation into else branch.